build_runner/lib/src/package_graph/apply_builders.dart

```diff
@@ -0,0 +1,61 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:build/build.dart';
+
+import '../generate/phase.dart';
+import 'package_graph.dart';
+
+bool _alwaysTrue(_) => true;
+
+bool Function(PackageNode) _dependencyFilter(String dependency) =>
+    (p) => p.dependencies.any((d) => d.name == dependency);
+
+bool Function(PackageNode) _isPackage(String package) =>
+    (p) => p.name == package;
+
+bool Function(PackageNode) _packagesContains(Set<String> packages) =>
+    (p) => packages.contains(p.name);
+
+class BuilderApplication {
+  final Builder builder;
+
+  /// Determines which package need [builder] applied.
+  final bool Function(PackageNode) filter;
+
+  /// Apply [builder] to all packages in the dependency graph.
+  const BuilderApplication.allPackage(this.builder) : filter = _alwaysTrue;
+
+  /// Apply [builder] to the pacakges which have a direct dependency on the
+  /// package named [dependency].
+  BuilderApplication.dependentsOf(this.builder, String dependency)
+      : filter = _dependencyFilter(dependency);
+
+  /// Apply [builder] to a single package in the graph.
+  BuilderApplication.singlePackage(this.builder, String package)
+      : filter = _isPackage(package);
+
+  /// Apply [builder] to a set of explicitly listed packages.
+  BuilderApplication.onPackages(this.builder, Set<String> packages)
+      : filter = _packagesContains(packages);
+}
```

Comment:
I'd like to hold off until we have a concrete use case. There's already a workaround of examining the `packageGraph` manually and dumping all relevant packages into a `Set`.

As I mentioned above I made the `_filter` field private and so if we can avoid exposing details of what this does I think the named constructors are the most descriptive way to use this.

Reply:
It is easy enough to add later, so I think I am fine w/this.

We will need a `BuilderApplication.transitiveDependentsOf` I think for angular - technically anything which transitively depends on angular might need codegen because of exports and provider lists. Right @matanlurey?